File: 3773-minimum-pair-removal-to-sort-array-i/minimum-pair-removal-to-sort-array-i.java
class Solution {
    public int minimumPairRemoval(int[] nums) {
        // Convert array to ArrayList for easy removal/insertion
        List<Integer> list = new ArrayList<>();
        for (int n : nums) list.add(n);

        int ops = 0;

        // Loop until the list is sorted
        while (!isSorted(list)) {
            int minSum = Integer.MAX_VALUE;
            int minIdx = -1;

            // Find the adjacent pair with the minimum sum
            for (int i = 0; i < list.size() - 1; i++) {
                int sum = list.get(i) + list.get(i + 1);
                
                // Strictly less than ensures we pick the leftmost in case of tie
                if (sum < minSum) {
                    minSum = sum;
                    minIdx = i;
                }
            }

            // Perform the merge operation
            int newVal = list.get(minIdx) + list.get(minIdx + 1);
            list.remove(minIdx);     // Remove first element of pair
            list.remove(minIdx);     // Remove second element (index shifts)
            list.add(minIdx, newVal); // Insert sum
            
            ops++;
        }
        return ops;
    }

    // Helper function to check if list is sorted
    private boolean isSorted(List<Integer> list) {
        for (int i = 0; i < list.size() - 1; i++) {
            if (list.get(i) > list.get(i + 1)) return false;
        }
        return true;
    }
}
  

File: 3773-minimum-pair-removal-to-sort-array-i/minimum-pair-removal-to-sort-array-i.cpp
class Solution {
public:
    int minimumPairRemoval(vector<int>& nums) {
        int ops = 0;
        while (true) {
            // Check if sorted
            bool sorted = true;
            for (int i = 0; i < nums.size() - 1; ++i) {
                if (nums[i] > nums[i+1]) {
                    sorted = false;
                    break;
                }
            }
            if (sorted) return ops;

            // Find min pair
            int minSum = INT_MAX;
            int idx = -1;
            for (int i = 0; i < nums.size() - 1; ++i) {
                long long currentSum = (long long)nums[i] + nums[i+1];
                if (currentSum < minSum) {
                    minSum = currentSum;
                    idx = i;
                }
            }

            // Replace pair with sum
            int newVal = nums[idx] + nums[idx+1];
            nums.erase(nums.begin() + idx);
            nums.erase(nums.begin() + idx);
            nums.insert(nums.begin() + idx, newVal);
            ops++;
        }
    }
};
  